cpp/src/arrow/device.h

```diff
@@ -98,6 +101,71 @@ class ARROW_EXPORT Device : public std::enable_shared_from_this<Device>,
   /// \brief Return the DeviceAllocationType of this device
   virtual DeviceAllocationType device_type() const = 0;
 
+  /// \brief EXPERIMENTAL: An object that provides event/stream sync primitives
+  class ARROW_EXPORT SyncEvent {
+   public:
+    virtual ~SyncEvent() = default;
+
+    /// @brief Block until sync event is completed.
+    virtual Status wait() = 0;
+
+    /// @brief Make the provided stream wait on the sync event.
+    ///
+    /// Tells the provided stream that it should wait until the
+    /// synchronization event is completed without blocking the CPU.
+    /// @param stream Should be appropriate for the underlying device
```

Comment:
Does this need to be exposed at the base class level if it takes a device-specific param?

Reply:
Ideally we could have a single API that allows people to write device-agnostic code where possible. Waiting on an event in a stream is a consistent concept across CUDA, HIP, SYCL, and OpenCL, but if possible we should try to make the API play nicely with CPUs as well.

Reply:
I'm not sure what waiting on a stream would mean for CPU.

Would this be more satisfactory/clear if instead of using bare `void*` here we also introduced `Device::Stream` to wrap each device's stream?

Reply:
I agree, where for CPU you could imagine it just taking a `nullptr` and being a no-op for ABI compatibility?

Reply:
Hmm, For a `Device::Stream` I guess we'd probably want that to be represented by passing around a `unique_ptr` of it? Or in the case here, maybe we take the stream via const ref and for a CPU the internal stream is null? I think at the moment the assumption is precisely what @kkraus14 suggested that for CPU it would just take a `nullptr` and be a no-op.

Reply:
There is no kind of explicit synchronization on CPUs (except if talking about memory barriers perhaps?), so it seems it would be a no-op instead.

Reply:
If the stream isn't owned by Arrow then a `unique_ptr` is not necessary. We could probably start with a `void*`.

Reply:
@pitrou Well it's already a `void*` currently, so I'm trying to determine what a `Device::Stream` would actually be wrapping, unless we're going to just do something like 

```c++
using Stream = void*;
...
Status StreamWait(Stream stream) = 0;
```

Otherwise is it really useful for it to be anything more than 

```c++
struct Stream {
    void* stream_;
};
```

Or am i missing something? I'm not sure if introducing a `Device::Stream` to wrap the stream is very useful in the long run, other than perhaps allowing a user to derive from it to provide more idiomatic and typed APIs?

Reply:
Well, @bkietz proposed `Stream` so he would have to explain what he meant by that exactly :-)

That said, the question of using a `void*` vs. a `Stream` abstraction is orthogonal to the question of using of using a raw pointer vs. a `unique_ptr`.

Reply:
I was thinking that `Device::Stream` would be an opaque and empty base class, mostly just a wrapper around `void*` to make it more clear that it's a Device-specific object. No ownership is necessary here since as @pitrou noted these won't be owned by arrow and users of this API will not want to maintain a `map<CUstream, std::shared_ptr<Device::Stream>>` in addition to their own stream pooling. Instead, subclasses of `Device::Stream` should be trivially constructible from their native counterparts.

```c++
class Device {
  // ...

  /// A synchronization stream in which SyncEvents may be sequenced for non
  /// host/CPU devices.
  class Stream {
   public:
    virtual const void* get_raw() const { return nullptr; }
   protected:
    Stream() = default;
  };

  class SyncEvent {
    // ...
    virtual Status StreamWait(const Stream&) = 0;   
    virtual Status Record(const Stream&) = 0;
  };
};

// CPU device can have its own placeholder stream, if necessary
const Stream& CpuDevice::stream() { ... }
```

A subclass would be pretty trivial:
```c++
class CudaDevice {
  // ...

  class Stream : public Device::Stream {
   public:
    const void* get_raw() const override { return &stream; }
    const CUstream stream;
    explicit Stream(CUstream stream) : stream{stream} {}

    // if devices have special stream values those can be specified here
    static Stream default_() { return Stream{0}; }
  };
};
```

... usage would just involve constructing the appropriate Stream whenever a sync event function is called:
```c++
CUstream stream = GetStreamFromPool();
// ...
event->StreamWait(CudaDevice::Stream{stream});
```

It's pretty trivial, but I think it wraps the intent into more OO-obvious terms

Reply:
Gotcha, that makes a lot more sense and I actually kinda like it. I'll quickly add that

Reply:
added in most recent commit